data_structures/linked_list/rotate_linked_list_by_k.py

```diff
@@ -0,0 +1,147 @@
+from __future__ import annotations
+
+
+class Node:
+    def __init__(self, data: int) -> None:
+        self.data = data
+        self.next = None
+
+
+def print_linked_list(head: Node | None) -> None:
+    """
+        Print the entire linked list iteratively.
+
+        This function prints the elements of a linked list separated by '->'.
+
+        Parameters:
+            head (Node | None): The head of the linked list to be printed,
+    or None if the linked list is empty.
+
+        >>> head = None
+        >>> head = insert_node(head, 0)
+        >>> head = insert_node(head, 2)
+        >>> head = insert_node(head, 1)
+        >>> print_linked_list(head)
+        0->2->1
+        >>> head = insert_node(head, 4)
+        >>> head = insert_node(head, 5)
+        >>> print_linked_list(head)
+        0->2->1->4->5
+    """
+    if head is None:
+        return
+    while head.next is not None:
+        print(head.data, end="->")
+        head = head.next
+    print(head.data)
+
+
+def insert_node(head: Node | None, data: int) -> Node:
+    """
+    Insert a new node at the end of a linked list and return the new head.
+
+    Parameters:
+        head (Node | None): The head of the linked list.
+        data (int): The data to be inserted into the new node.
+
+    Returns:
+        Node: The new head of the linked list.
+
+    >>> head = None
+    >>> head = insert_node(head, 10)
+    >>> head = insert_node(head, 9)
+    >>> head = insert_node(head, 8)
+    >>> print_linked_list(head)
+    10->9->8
+    """
+    new_node = Node(data)
+    # If the linked list is empty, the new_node becomes the head
+    if head is None:
+        return new_node
+
+    temp_node = head
+    while temp_node.next is not None:
+        temp_node = temp_node.next
+
+    temp_node.next = new_node  # type: ignore
+    return head
+
+
+def right_rotate_by_k(head: Node | None, rotation: int) -> Node | None:
+    """
+    Rotate a linked list to the right by rotation times.
+
+    Parameters:
+        head (Node | None): The head of the linked list.
+        rotation (int): The number of places to rotate.
+
+    Returns:
+        Node | None: The head of the rotated linked list\
+if the linked list is not None, otherwise None.
```

Comment:
[Mypy](mypy.readthedocs.io) can test data types in the function signature, but not in the comments.

Do not repeat the datatypes in both places because readers will be confused if one is changed and the other is not.

How about this naming...
```suggestion
def rotate_to_the_right(head: Node | None, places: int) -> Node | None:
    """
    Rotate a linked list to the right by a number of places.

    Parameters:
        head: The head of the linked list.
        places: The number of places to rotate the linked list

    Returns:
        The head of the rotated linked list or None if there is no list.
```

Reply:
I might even raise a `ValueError` if I was handed a head=None.

Reply:
Do we have a test for a linked list that has just 3 elements but we are asked to shift to the right 10 places?

Reply:
What happens if we are asked to shift to the right -2 places?  -2.5 places?

Reply:
> What happens if we are asked to shift to the right -2 places? -2.5 places?

If the rotation number is not an integer, it should raise an error. If it is a negative integer, converting it to a positive integer would be a better idea, right?

Reply:
> Do we have a test for a linked list that has just 3 elements but we are asked to shift to the right 10 places?

Yes, I tested it, and it is working correctly.